[BOLT] Add GNUPropertyRewriter and warn on AArch64 BTI note

bgergely0 opened the pull request:
This commit adds the GNUPropertyRewriter, which parses features from the .note.gnu.property section.

Currently we only read the bit indicating BTI support (GNU_PROPERTY_AARCH64_FEATURE_1_BTI).

As BOLT does not add BTI landing pads to targets of indirect branches/calls, we have to emit a warning that the output binary may be corrupted.

bgergely0 committed:
[BOLT] Add GNUPropertyRewriter and warn on AArch64 BTI note

This commit adds the GNUPropertyRewriter, which parses features from the
.gnu.property.note section.

Currently we only read the bit indicating BTI support
(GNU_PROPERTY_AARCH64_FEATURE_1_BTI).

As BOLT does not add BTI landing pads to targets of indirect
branches/calls, we have to emit a warning that the output binary may be
corrupted.

paschalis-mpeis reviewed (approved):
Hey Gergely,

Thanks for adding support for reading features from `.note.gnu.property`.
I added a few nits.

bgergely0 committed:
[BOLT] Review changes

- full warning message in lit tests
- \p in comment
- more specific error messages

bgergely0 commented:
Thanks for the review and approval @paschalis-mpeis! I addressed the nits.

I will wait a few days before merging to give some time for objections.

bgergely0 committed:
[BOLT] correct name of .note.gnu.property section

.gnu.property.note -> .note.gnu.property
